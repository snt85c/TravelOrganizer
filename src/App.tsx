import { UserAuthContextProvider } from "./LoginComponents/UserAuth";
import "./App.css";
import Main from "./Main";
import { HashRouter } from "react-router-dom";

const App: React.FC = () => {
  return (
    <HashRouter>
      <UserAuthContextProvider>
        <Main />
      </UserAuthContextProvider>
    </HashRouter>
  );
};

export default App;
