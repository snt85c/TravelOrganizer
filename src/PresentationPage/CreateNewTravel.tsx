import { useState } from "react";
import { arrayUnion, doc, updateDoc } from "firebase/firestore";
import { db } from "../LoginComponents/firebase";
import { iTravel } from "../Interface";

export default function CreateNewTravel(props: {
  travelList: [iTravel?];
  setTravelList: React.Dispatch<React.SetStateAction<[iTravel?]>>;
}) {
  const [newTravel, setNewTravel] = useState<string>("");

  const handleClick = async () => {
    const tempTravelList: any = props.travelList.slice(); //pass by value, not reference by slice function that returns a new array
    const newTravelObject: iTravel = { name: newTravel, id: Date.now() };
    tempTravelList.push(newTravelObject);
    props.setTravelList(tempTravelList);
    // console.log(tempTravelList === props.travelList,  "equality add")
    //add to cloud firestore
    try {
      await updateDoc(doc(db, "travels", "NTyNtjKvHwnEcbaOI73f"), {
        travel: arrayUnion(newTravelObject),
      });
    } catch (e) {
      console.log(e);
    }
  };
  return (
    <>
      <div className="flex flex-col justify-center items-center m-5">
        <input
          className="rounded shadow-md text-black text-center"
          value={newTravel}
          onChange={(e) => setNewTravel(e.target.value)}
        />
        <div onClick={handleClick}>
          {!newTravel ? "new travel" : "add " + newTravel}
        </div>
      </div>
    </>
  );
}
