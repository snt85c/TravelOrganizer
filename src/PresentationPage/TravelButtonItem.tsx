import {
  collection,
  doc,
  DocumentData,
  getDoc,
  getDocs,
  setDoc,
  updateDoc,
} from "firebase/firestore";
import { ImEnter, ImEye } from "react-icons/im";
import { useEffect, useRef, useState } from "react";
import { iTravel, iTravelData } from "../Interface";
import { db } from "../LoginComponents/firebase";
import { telegramBotKey, chat_id } from "../Main";
import {  useNavigate } from "react-router-dom";

export default function TravelButtonItem(props: {
  i?: number;
  data?: iTravel;
  loggedUser: any;
  setTravel: React.Dispatch<React.SetStateAction<iTravel>>;
  travelList: [iTravel?];
  setTravelList: React.Dispatch<React.SetStateAction<[iTravel?]>>;
  setIsWatching:React.Dispatch<React.SetStateAction<boolean>>,
  watchTravel: Function;
  joinTravel: Function;
}) {
  const [isEditing, setIsEditing] = useState<boolean>(false);
  const [isRenaming, setIsRenaming] = useState<boolean>(false);
  const [isDeleting, setIsDeleting] = useState<boolean>(false);
  const [newName, setNewName] = useState<string>("");
  const { ref } = HandleClickOutsideComponent(setIsEditing);
  const navigate = useNavigate();


  function isAuthor() {
    return props.loggedUser?.uid === props.data?.createdBy;
  }

  function HandleClickOutsideComponent(
    setState: React.Dispatch<React.SetStateAction<boolean>>
  ) {
    const ref = useRef<HTMLDivElement>(null);

    const handleClickOutside = (event: Event) => {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        setState(false);
      }
    };

    useEffect(() => {
      document.addEventListener("click", handleClickOutside, true);
      return () => {
        document.removeEventListener("click", handleClickOutside, true);
      };
    });

    return { ref };
  }

  function telegramAlertDeleteTravel() {
    const data = `travel: ${props.data?.name} has been deleted by: ${
      props.loggedUser.displayName.split(" ")[0]
    }`;

    fetch(
      `https://api.telegram.org/bot${telegramBotKey}/sendMessage?chat_id=${chat_id}&text=${data} `
    );
  }

  const handleDelete = async () => {
    setIsDeleting(false);
    async function handleDeleteOnFirestore(userToUpdate: iTravelData) {
      await setDoc(doc(db, "users", userToUpdate.userInfo.uid), userToUpdate, {
        merge: false,
      });
    }
    try {
      let travelIdToDelete: number = props.data?.id ? props.data?.id : 0;
      let filteredTravelList: any = props.travelList?.filter((item) => {
        return item?.id !== props.data?.id;
      });

      props.setTravelList(filteredTravelList);
      await updateDoc(doc(db, "travels", "NTyNtjKvHwnEcbaOI73f"), {
        travel: filteredTravelList,
      });

      const querySnapshot = await getDocs(collection(db, "users"));
      querySnapshot.forEach((doc) => {
        //get each user and if there is a key with the travelIdToDelete, delete the key then update firestore
        let currentUser: iTravelData = doc.data() as iTravelData;
        if (currentUser[travelIdToDelete]) {
          delete currentUser[travelIdToDelete];
          handleDeleteOnFirestore(currentUser);
        }
      });
      telegramAlertDeleteTravel();
      props.setTravel({ name: "", id: 0, createdBy: "", userName: "" });
    } catch (e) {
      console.log(e);
    }
  };

  const handleRename = async () => {
    setIsRenaming(false);
    setIsDeleting(false);
    setIsEditing(false);
    try {
      let travelIdToRename: number = props.data?.id ? props.data?.id : 0;
      let newNameForSelectedTravel: string = newName;
      const updateUserOnFirestore = async (uid: string, user: iTravelData) => {
        await setDoc(doc(db, "users", uid), user, { merge: true });
      };

      //GET AND MODIFY EACH USERS TO CHANGE THE TRAVEL NAME
      const querySnapshot = await getDocs(collection(db, "users"));
      querySnapshot.forEach((doc) => {
        let currentUser: iTravelData = doc.data() as iTravelData;
        if (currentUser[travelIdToRename]) {
          currentUser[travelIdToRename].name = newNameForSelectedTravel;
          updateUserOnFirestore(
            currentUser[travelIdToRename].userInfo.uid,
            currentUser
          );
        }
      });

      //GET AND MODIFY TRAVELS FROM FIRESTORE
      const querySnapshotTravels = await getDoc(
        doc(db, "travels", "NTyNtjKvHwnEcbaOI73f")
      );
      let travelsFromFirestore: DocumentData | { travel: [] } | undefined =
        querySnapshotTravels.data();
      travelsFromFirestore?.travel.forEach(
        (temp: { id: number; name: string }) => {
          if (temp.id === travelIdToRename) {
            temp.name = newNameForSelectedTravel;
          }
        }
      );
      await updateDoc(doc(db, "travels", "NTyNtjKvHwnEcbaOI73f"), {
        ...travelsFromFirestore,
      });

      //GET AND MODIFY TRAVELLIST STATE
      let tempArray: [(iTravel | undefined)?] = [...props.travelList];
      tempArray.forEach((item) => {
        if (item?.id === travelIdToRename) {
          item.name = newNameForSelectedTravel;
        }
      });
      props.setTravelList(tempArray);
    } catch (e) {
      console.log(e);
    }
  };

  const handleClickSetTravel = () => {
    const temp: iTravel = props.data?.name
      ? props.data
      : { name: "", id: 0, createdBy: "", userName: "" };
    props.setTravel(temp);
  };

  const handleEdit = () => {
    setIsEditing(!isEditing);
    setIsRenaming(false);
    setIsDeleting(false);
  };



  return (
    <div
      className="flex flex-col relative w-[1/4] m-1 p-1 mx-10 md:mx-40 justify-center items-center text-black rounded bg-white border duration-300 "
      style={{
        height: isEditing
          ? isRenaming || isDeleting
            ? "130px"
            : "100px"
          : "70px",
        justifyContent: isEditing ? "space-evenly" : "center",
      }}
    >
      <div className="flex justify-between w-full">
        <div
          onClick={() => {
            handleClickSetTravel();
            props.setIsWatching(true)
          }}
          className="m-2 flex flex-col items-center justify-center select-none cursor-pointer"
        >
          {
            <>
              <div>
                {" "}
                <ImEye size={20} />
              </div>
              <div className="-mt-2">View</div>{" "}
            </>
          }
        </div>
        <div className="flex flex-col justify-center items-center">
          <div className="text-[2.7vw] sm:text-[0.9rem] mt-1 text-gray-800  select-none font-[homeworld-norm]"
          >
            {props.data?.name.toUpperCase()}
          </div>
          <div className="text-[0.5rem] -my-1 select-none">
            created by:{" "}
            <span className="text-[0.7rem] text-pink-600 font-bold">
              {isAuthor() ? "You" : props.data?.userName}
            </span>{" "}
            id:
            <span className="text-[0.7rem] text-pink-600 font-bold">
              {props.data?.id}
            </span>
          </div>
        </div>
        <div
          onClick={() => {
            handleClickSetTravel();
            props.setIsWatching(false)
            navigate("/user")
          }}
          className="m-2 flex flex-col items-center justify-center select-none cursor-pointer"
        >
          {props.loggedUser && 
            <>
              <div>
                {" "}
                <ImEnter size={20} />
              </div>
              <div className="-mt-2">Join</div>
            </>
          }
        </div>
      </div>
      {isAuthor() && (
        <div className="flex -mt-1">
          {!isEditing && (
            <div
              className="mx-2 mb-1 text-sm cursor-pointer text-gray-800 hover:text-amber-500 duration-300 select-none"
              onClick={handleEdit}
            >
              Edit
            </div>
          )}
          {isEditing && (
            <div ref={ref}>
              <div className="flex flex-col justify-evenly items-center">
                <div className="flex flex-row p-1 m-1">
                  <div
                    className="mx-2  text-sm cursor-pointer text-gray-800 hover:text-amber-500 duration-300 select-none"
                    onClick={() => {
                      setIsDeleting(!isDeleting);
                      setIsRenaming(false);
                    }}
                  >
                    Delete
                  </div>
                  <div
                    className="mx-2 text-sm cursor-pointer text-gray-800 hover:text-amber-500 duration-300 select-none"
                    onClick={() => {
                      setIsRenaming(!isRenaming);
                      setIsDeleting(false);
                    }}
                  >
                    Rename
                  </div>
                </div>
                {isRenaming && (
                  <div
                    onKeyDown={(e) => {
                      if (e.key === "Enter" && isRenaming) handleRename();
                    }}
                  >
                    <input
                      className="rounded-xl border-2 border-amber-500 mx-2 px-2 text-center"
                      onChange={(e) => setNewName(e.target.value)}
                    />
                    <button onClick={handleRename}>ok</button>
                  </div>
                )}
                {isDeleting && (
                  <div
                    className=" flex flex-col justify-center items-center mx-2  text-sm cursor-pointer text-gray-800 hover:text-red-600 duration-300 hover:font-bold select-none"
                    onClick={handleDelete}
                  >
                    <div>press to delete</div>
                    <div className="text-[0.5rem] -mt-2">
                      this will cancel your data permanently
                    </div>
                  </div>
                )}
              </div>
            </div>
          )}
        </div>
      )}
    </div>
  );
}
