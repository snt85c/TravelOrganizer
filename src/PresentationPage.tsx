import flairImage from "./img/undraw_explore_re_8l4v.svg";
import flairImage2 from "./img/undraw_journey_re_ec5q.svg";
import flairImage3 from "./img/undraw_travelers_re_y25a.svg";
import TravelButtonItem from "./TravelButtons";

export default function PresentationPage(props: {
  travels: any;
  setTravel: React.Dispatch<React.SetStateAction<string>>;
}) {
  const travelButtonsList =
    props.travels &&
    props.travels[0].map((travel: string, i: number) => {
      return <TravelButtonItem key={i} data={travel} setTravel={props.setTravel}/>;
    });

  return (
    <div className="flex flex-col my-5 justify-between h-[80%]">
      <div className="flex my-5 justify-center items-center font-[homeworld-bold]">
        BACKPACK ORGANIZER
      </div>
      {/* <div className="flex flex-col w-[1/4] p-2">
        {props.travels && travelButtonsList}
      </div> */}
      <img
        src={flairImage3}
        style={{
          backgroundImage:
            "radial-gradient(ellipse, rgb(100 116 139), transparent 70%)",
        }}
        className="absolute h-[200px] bottom-10 right-10 "
      />
    </div>
  );
}
