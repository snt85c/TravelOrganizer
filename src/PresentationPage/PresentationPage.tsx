import flairImage from "../img/undraw_travelers_re_y25a.svg";
import TravelButtonItem from "./TravelButtonItem";
import CreateNewTravel from "./CreateNewTravel";
import { iTravel } from "../Interface";

export default function PresentationPage(props: {
  travelList: [iTravel?];
  setTravel: React.Dispatch<React.SetStateAction<String>>;
  setTravelList: React.Dispatch<React.SetStateAction<[iTravel?]>>;
}) {
  const travelButtonsList = props.travelList.map(
    (currentData?: iTravel, i?: number) => {
      return (
        <TravelButtonItem
          key={i}
          i={i}
          data={currentData}
          setTravel={props.setTravel}
          travelList={props.travelList}
          setTravelList={props.setTravelList}
        />
      );
    }
  );

  return (
    <div className="flex flex-col my-5 justify-between h-[80%]">
      <div className="flex my-5 justify-center items-center font-[homeworld-bold]">
        BACKPACK ORGANIZER
      </div>
      <div className="flex flex-col w-[1/4] p-2">
        <div>{travelButtonsList}</div>
        <CreateNewTravel
          travelList={props.travelList}
          setTravelList={props.setTravelList}
        />
      </div>
      <img
        src={flairImage}
        style={{
          backgroundImage:
            "radial-gradient(ellipse, rgb(100 116 139), transparent 70%)",
        }}
        className="absolute h-[100px] md:h-[200px] bottom-10 right-5 md:right-10 "
      />
    </div>
  );
}
