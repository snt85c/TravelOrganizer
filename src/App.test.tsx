import { getByTestId, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import PresentationPage from "./PresentationPageComponent/PresentationPage";

// test("renders learn react link", () => {
//   render(<App />);
//   const linkElement1 = screen.getByText(/Travel/i);
//   const linkElement2 = screen.getByText(/Organizer/i);
//   const linkElement3 = screen.getByText(/By Snt/i);

//   expect(linkElement1).toBeInTheDocument();
//   expect(linkElement2).toBeInTheDocument();
//   expect(linkElement3).toBeInTheDocument();
// });

// describe("App component", () => {
//   it("renders Apps", () => {
//     const { container } = render(<App />);
//     expect(container).toMatchSnapshot();
//   });
// });

// test("on render, check if we have 5 travel buttons", () => {
//   render(<App />)
//   const input = screen.getAllByTestId("travelButton")
//   expect(input).toHaveLength(5)
// })
