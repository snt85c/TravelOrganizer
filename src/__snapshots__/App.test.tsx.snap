// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders Apps 1`] = `
<div>
  <div
    class="bg-gray-800 relative pt-[60px] pb-5 min-h-full text-white"
  >
    <div
      class="absolute flex top-0 z-40 min-w-full px-2 md:px-20 py-2 bg-gray-800  justify-between items-center"
    >
      <div
        class="flex flex-col h-10 w-10 cursor-pointer"
      >
        <svg
          baseProfile="tiny"
          fill="currentColor"
          height="40"
          stroke="currentColor"
          stroke-width="0"
          version="1.2"
          viewBox="0 0 24 24"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22.262 10.468c-3.39-2.854-9.546-8.171-9.607-8.225l-.655-.563-.652.563c-.062.053-6.221 5.368-9.66 8.248-.438.394-.688.945-.688 1.509 0 1.104.896 2 2 2h1v6c0 1.104.896 2 2 2h12c1.104 0 2-.896 2-2v-6h1c1.104 0 2-.896 2-2 0-.598-.275-1.161-.738-1.532zm-8.262 9.532h-4v-5h4v5zm4-8l.002 8h-3.002v-6h-6v6h-3v-8h-3.001c2.765-2.312 7.315-6.227 9.001-7.68 1.686 1.453 6.234 5.367 9 7.681l-3-.001z"
          />
        </svg>
      </div>
      <div />
    </div>
    <div
      class="flex flex-row min-h-[2.3rem] justify-between mx-2 md:mx-20 select-none"
      style="opacity: 0; transform: translateY(-50px) translateZ(0);"
    />
    <div
      class="flex flex-col my-5 justify-between h-[80%]"
    >
      <div
        class="flex flex-col justify-center items-center min-h-[50vh]"
      >
        <div
          class=" text-[2rem] animate-pulse font-[homeworld-norm]"
        >
          LOADING
        </div>
      </div>
      <div
        class=" absolute z-0 bottom-7 left-2  font-[phonk] leading-none select-none"
      >
        <div
          class="flex text-[20vw] sm:text-[8rem]"
        >
          Travel
        </div>
        <div
          class="text-[10vw] sm:text-[4rem] -mt-[3vh]"
        >
          organizer
          <span
            class="text-[3vw] sm:text-[1rem] -mt-[3vh]"
          >
            by Snt
          </span>
        </div>
      </div>
      <img
        class="absolute z-0 h-[100px] hidden sm:flex md:h-[200px] bottom-10 right-5 md:right-10 "
        src="undraw_travelers_re_y25a.svg"
      />
    </div>
    <div
      class="absolute select-none flex justify-between items-center h-5 px-5 text-[0.8rem] text-gray-400 bg-gray-800 min-w-full bottom-0 z-20"
    >
      <div>
        Snt85c
      </div>
      <a
        href="https://github.com/snt85c"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 496 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
